drivers/arm/ws2812.c

```diff
@@ -0,0 +1,111 @@
+#include "quantum.h"
+#include "ws2812.h"
+#include "ch.h"
+#include "hal.h"
+
+#if !defined(LED_ARRAY) && defined(RGB_MATRIX_ENABLE)
+// LED color buffer
+LED_TYPE led[DRIVER_LED_TOTAL];
+#    define LED_ARRAY led
+#endif
+
+#ifdef RGB_MATRIX_ENABLE
+// Set an led in the buffer to a color
+void inline ws2812_setled(int i, uint8_t r, uint8_t g, uint8_t b) {
+    led[i].r = r;
+    led[i].g = g;
+    led[i].b = b;
+}
+
+void ws2812_setled_all(uint8_t r, uint8_t g, uint8_t b) {
+    for (int i = 0; i < sizeof(led) / sizeof(led[0]); i++) {
+        led[i].r = r;
+        led[i].g = g;
+        led[i].b = b;
+    }
+}
+#endif
+
+/* adapted from
+https://github.com/bigjosh/SimpleNeoPixelDemo/blob/master/SimpleNeopixelDemo/SimpleNeopixelDemo.ino
+*/
+
+#define NS_PER_SEC (1000000000L)  // Note that this has to be SIGNED since we want to be able to check for negative values of derivatives
+
+#define CYCLES_PER_INSTRUCTION 0.35
+
+#define CYCLES_PER_SEC (STM32_SYSCLK * CYCLES_PER_INSTRUCTION)
+
+#define NS_PER_CYCLE (NS_PER_SEC / CYCLES_PER_SEC)
+
+#define NS_TO_CYCLES(n) ((n) / NS_PER_CYCLE)
+
+#define wait_ns(x)                              \
+    for (int i = 0; i < NS_TO_CYCLES(x); i++) { \
+        __NOP();                                \
+    }
+
+// These are the timing constraints taken mostly from the WS2812 datasheets
+// These are chosen to be conservative and avoid problems rather than for maximum throughput
+
+#define T1H 900  // Width of a 1 bit in ns
+#define T1L 600  // Width of a 1 bit in ns
+
+#define T0H 400  // Width of a 0 bit in ns
+#define T0L 900  // Width of a 0 bit in ns
```

Comment:
enclose these in `#ifndef` blocks so timing can be overridden for other chips -- 2020's are 300/800 timing

Reply:
Ive matched the timings to the AVR implementation, but as the protocol range is quite large I think it should be fine (though i don't have anything to test that against).

Reply:
For WS2812(B|C)-2020:
T1H = 5800-1000
T1L = 220-420
T0H = 220-380
T0L = 580-1000